### ISSUE TITLE
onRegionChangeComplete props Callback triggering infinitely without any user input on Android.

### ISSUE THREAD

**Issue by akshay2604** [react-community/react-native-maps#846]
i am not using fitToSuppliedMarkers as another issue mentioned, only plain MapComponent . For some reason on Android onRegionChangeComplete is being called infinitely many times. However it is working correctly on IOS.

React native - 0.35
react-native-maps - 0.10.4

**Comment by akshay2604** [comment]
if you use region prop along with onRegionChangeComplete , the onRegionChangeComplete is called as a callback. onRegionChangeComplete is called whenever onChange event occurs natively. it could be due to user dragging map, animation on map or even setting up of gps location. If possible another prop should be added like onPanDragComplete so that we get callback only when user drags map and finishes dragging it.

**Comment by M3po** [comment]
Did you make it work and how? i am also having problem with infinite loop

**Comment by akshay2604** [comment]
`<MapView
          ref={(ref) => { this.map = ref; }}
          style={styles.map}
          initialRegion={this.props.initialRegion}
          loadingEnabled
          onRegionChangeComplete={(region) => {
            // do something
            }
          }}
        >
</MapView>
`
This is my new mapComponent. It turns out i was using region prop and was only setting state of region. removing it also works.
I don't know how to use both together. If you look at components/MapView.js there is a particular function
`_onChange(event) {
    this.__lastRegion = event.nativeEvent.region;
    if (event.nativeEvent.continuous) {
      if (this.props.onRegionChange) {
        this.props.onRegionChange(event.nativeEvent.region);
      }
    } else if (this.props.onRegionChangeComplete) {
      this.props.onRegionChangeComplete(event.nativeEvent.region);
    }
  }`

The _onChange method is called whenever any event happens.If whatever you write in onRegionChangeComplete **triggers a call to region prop** , region prop in itself triggers an _onChange event which causes onRegionChangeComplete to call. This quickly spirals to an infinite loop each callback trigering each other. You can put logs in MapView.js file to see it for yourself.
 
I think something like a panDragComplete in this case would do the trick. I am not sure if mapkit or google maps provider provide a native event for this or not. But until then make sure onRegionChangeComplete has code without any side effects. Or if you can't do that, remove region prop, it worked for me

**Comment by M3po** [comment]
Thanks for the input i will try to edit the component so i might get something (or will at the do without region change until they make something like you suggested). You should leave this issue open.

**Comment by meddy** [comment]
I have an app flow like so:
1. user enters map scene
2. user's location is passed to region prop of MapView component
3. user's location is updated
4. new location is passed to region prop of MapView component
5. when onRegionChangeComplete is executed, dispatch some actions to update the state, so we can update some markers on the map

On iOS this works normally. OnRegionChangeComplete gets called twice, as you would expect.

On android, as @akshay2604 mentioned, if a re-render or state change happens during an onRegionChangeComplete callback, then onRegionChangeComplete callbacks get called repeatedly causing an infinite loop.

In order to work around this for now I implemented a check in my onRegionChangeComplete callback to determine if the region has actually changed. If not, return early instead of dispatching actions. This at least eliminates multiple HTTP requests or re-renders occurring.

Example Code:

```
<MapView
        style={style.map}
        onRegionChangeComplete={(region) => {
          const {initialRegion} = this.state;
          
          if (initialRegion && isEqual(initialRegion, region)) {
            return;
          }

          this.setState({initialRegion: region});
          
          // dispatch some actions
        }}
/>
```

**Comment by heikkihautala** [comment]
Using version `0.15.2`, I solved this problem by having 2 different region props in the state: `mapRegion` and `setMapRegion`.

`mapRegion`:
- Used when doing stuff when ever map region changes.
- Gets updated in `onRegionChange` and `onRegionChangeComplete`. 

`setMapRegion`:
- Is bound to the MapView's `region` property.
- Is set when ever I want update the region manually.

My MapView component looks like this:
```
<MapView
  region={this.state.setMapRegion}
  onRegionChange={(region) => {
      this.setState({
          mapRegion: region,
      })
      ...
  }}
  onRegionChangeComplete={(region) => {
      this.setState({
          mapRegion: region,
      })
      ...
  }}
/>
```

**Comment by paulsc** [comment]
I just had the same problem i.e.:

- I have a button that changes the map position by changing the "region" state variable
- The "region" state variable is bound to the region property of MapView
- so clicking the button triggers a render, which triggers onRegionChangeComplete()
- onRegionChangeComplete changes some other state variable
- which triggers a render() call
- which triggers an onRegionChangeComplete() call again => infinite loop

and what fixed it was setting the "region" state variable in my onRegionChangeComplete() callback. I am now setting it to the argument being passed to onRegionChangeComplete(). 

I'm not entirely sure why this fixes it, but I assume it has something to do with the state being the same so it doesn't trigger additional render() calls?

I'm curious if this works for anyone else. 

Cheers

**Comment by augisticz** [comment]
I'm use `onRegionChangeComplete={e => this.setState({ region: e })}` is work

**Comment by IjzerenHein** [comment]
Should be fixed by PR #1710

**Comment by alvelig** [comment]
I discourage controlling region state. Try this: https://github.com/react-community/react-native-maps/issues/209#issuecomment-350907665

Anyway it should be fixed with the PR above.

**Comment by tsangwailam** [comment]
I found that when you zoom to max. The region from onRegionChangeComplete will be continue changing. The change is so small and about 15/16 significant decimal digits. Is it related to the permission?

**Comment by mskuser1** [comment]
@tsangwailam did u find a solution? I have the same problem, it only moves by itself when zoom is on max

**Comment by iamshadmirza** [comment]
> 
> 
> ```
> <MapView
>           ref={(ref) => { this.map = ref; }}
>           style={styles.map}
>           initialRegion={this.props.initialRegion}
>           loadingEnabled
>           onRegionChangeComplete={(region) => {
>             // do something 
>             }
>           }}
>         >
> </MapView>
> ```
> 
> This is my new mapComponent. It turns out i was using region prop and was only setting state of region. removing it also works.
> I don't know how to use both together. If you look at `components/MapView.js` there is a particular function
> 
> ```
> _onChange(event) {
>     this.__lastRegion = event.nativeEvent.region;
>     if (event.nativeEvent.continuous) {
>       if (this.props.onRegionChange) {
>         this.props.onRegionChange(event.nativeEvent.region);
>       }
>     } else if (this.props.onRegionChangeComplete) {
>       this.props.onRegionChangeComplete(event.nativeEvent.region);
>     }
>   }
> ```
> 
> The _onChange method is called whenever any event happens.If whatever you write in onRegionChangeComplete **triggers a call to region prop** , region prop in itself triggers an _onChange event which causes onRegionChangeComplete to call. This quickly spirals to an infinite loop each callback trigering each other. You can put logs in MapView.js file to see it for yourself.
> 
> I think something like a panDragComplete in this case would do the trick. I am not sure if mapkit or google maps provider provide a native event for this or not. But until then make sure onRegionChangeComplete has code without any side effects. Or if you can't do that, remove region prop, it worked for me

You saved my day

**Comment by yasir-netlinks** [comment]
Hi guys, any solution for this issue?

**Comment by Pratit23** [comment]
> > if you use region prop along with onRegionChangeComplete , the onRegionChangeComplete is called as a callback. onRegionChangeComplete is called whenever onChange event occurs natively. it could be due to user dragging map, animation on map or even setting up of gps location. If possible another prop should be added like onPanDragComplete so that we get callback only when user drags map and finishes dragging it.
> 
> Yes, you're right. Adding `initialRegion` instead of region solves the problem. However, there is one issue, after `onRegionChangeComplete` is done saving new region, the map is not updating with the new region we just saved. That only works with `region` prop. Is there any solution to this issue?

Did you get a fix for this?

**Comment by JimTeva** [comment]
For anyone using GoogleMaps, this is how I fixed it (version 1.2.0):
```
region={this.state.region}
onRegionChangeComplete={(region, gesture) => {
  if (gesture.isGesture) {
    this.setState({ region });
  }
}}
```
note: gesture.isGesture indicates if the change was made by user or animation.

**Comment by zubbey** [comment]
> For anyone using GoogleMaps, this is how I fixed it (version 1.2.0):
> 
> ```
> region={this.state.region}
> onRegionChangeComplete={(region, gesture) => {
>   if (gesture.isGesture) {
>     this.setState({ region });
>   }
> }}
> ```
> 
> note: gesture.isGesture indicates if the change was made by user or animation.

Thanks
